[HUDI-271] Create QuickstartUtils class towards simplifying quickstar…

bhasudha opened the pull request:
…t guide

This class is a thin version of HoodieTestDataGenerator and will be used in Quickstart guide(Doc changes to follow in a separate PR). The intention is to simplify quickstart to showcase hudi APIs by writing and reading using spark datasources.
This is located in hudi-spark module intentionally to bring all the necessary classes in hudi-spark-bundle finally.

bvaradar commented on hudi-spark/src/main/java/org/apache/hudi/QuickstartUtils.java:
```diff
@@ -0,0 +1,319 @@
+  /**
+   * Example row change event based on some example data used by quickstart guide.
+   */
+  static class TestRawTripPayload implements HoodieRecordPayload<TestRawTripPayload> {
```
I think this class is implementing  HoodieRecordPayload  directly due to historical reasons. There is a generic OverwriteWithLatestAvroPayload which you can reuse.   OverwriteWithLatestAvroPayload keeps it as Avro record and avoids json. 
If you need to showcase soft deletes, you can have this class extend OverwriteWithLatestAvroPayload, have "is_deleted" field part of the schema and only override getInsertValue() to check if this flag is enabled after decoding and return accordingly.

That ways, users who look at this class do not feel like they have to write a lot of code.

bhasudha replied:
This is neat. Thanks for suggesting this. I am planning to showcase delete in next iteration. So for now, I ll just reuse OverwriteWithLatestAvroPayload.

bhasudha commented on hudi-spark/src/main/java/org/apache/hudi/QuickstartUtils.java:
```diff
@@ -0,0 +1,319 @@
+      this.existingKeys = keyPartitionMap;
+    }
+
+    private static String generateRandomCommitMsg() {
```
done!

bhasudha commented on hudi-spark/src/main/java/org/apache/hudi/QuickstartUtils.java:
```diff
@@ -0,0 +1,319 @@
+
+    public static class KeyPartition implements Serializable {
+      HoodieKey key;
+      String partitionPath;
```
sure.

bhasudha commented on hudi-spark/src/main/java/org/apache/hudi/QuickstartUtils.java:
```diff
@@ -0,0 +1,319 @@
+     * Generates a new avro record of the above schema format, retaining the key if optionally provided.
+     */
+    public static TestRawTripPayload generateRandomValue(HoodieKey key, String commitTime) throws IOException {
+      GenericRecord rec = generateGenericRecord(key.getRecordKey(), "rider-" + commitTime, "driver-" + commitTime, 0.0);
```
sure. Renamed to riderDriverSuffix

bhasudha commented:
@bvaradar I removed the KeyPartition static class and replaced with hoodieKey. I renamed all occurrences of commitTime with randomString or something along that line. Also, was able to reuse the OverwriteWithLatestAvroPayload class. Please review at your convenience.

bvaradar commented on hudi-spark/src/main/java/org/apache/hudi/QuickstartUtils.java:
```diff
@@ -0,0 +1,208 @@
+    }
+  }
+
+  public static List<String> convertToStringList(List<HoodieRecord> records) {
```
Is this method to show  the generated records in the demo ?

bhasudha replied:
@bvaradar yes. The idea is to present like below for insert:

val dataGen = new DataGenerator
val inserts = convertToStringList(dataGen.generateInserts(50)) 
val ds = spark.read.json(spark.sparkContext.parallelize(inserts, 2));
ds.write.format("org.apache.hudi").options(getQuickstartWriteConfigs).option(PRECOMBINE_FIELD_OPT_KEY, "ts").option(RECORDKEY_FIELD_OPT_KEY, "uuid").option(PARTITIONPATH_FIELD_OPT_KEY, "partitionpath").option(TABLE_NAME, tableName).mode(SaveMode.Append).save(basepath);

bvaradar commented:
Thanks @bhasudha for getting it done in short notice.

bhasudha commented on hudi-spark/src/main/java/org/apache/hudi/QuickstartUtils.java:
```diff
@@ -0,0 +1,208 @@
+public class QuickstartUtils {
+
+  public static class DataGenerator {
+    private static final String DEFAULT_FIRST_PARTITION_PATH = "2019/09/15";
```
No specific reason. I ll change it to latest.

vinothchandar replied:
Actually, can we name the partitions based on region/country/cities? 


Americas/US/SanFrancisco
Europe/UK/London
Asia/India/Chennai

This way its relevant even a year later ?

bhasudha replied:
yeah makes sense. I ll change the sample schema and update the PR.

bhasudha commented:
> @bhasudha can you please change the commit message to be bullets like the standard format

Done!
